### PR TITLE
ARROW-15896: [Python][C++] Add errno detail for filesystem "file not found" errors

### PR DESCRIPTION
**Pull request by pitrou**
This allows raising FileNotFoundError in Python, which is more user-friendly.

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit *hdfs*

**Review by coryan — APPROVED** [review]
The GCS changes LGTM.

File: cpp/src/arrow/util/io_util.cc [hunk 1]
```diff
@@ -230,6 +230,31 @@ class ErrnoDetail : public StatusDetail {
 #if _WIN32
 const char kWinErrorDetailTypeId[] = "arrow::WinErrorDetail";
 
+// Map from a Windows error code to a `errno` value
+int WinErrorToErrno(int errnum) {
+  // NOTE there's a more complete mapping in the UCRT source code
+  // (in misc/errno.cpp) but it's not under an open source license
```
Comment (pitrou) [inline cpp/src/arrow/util/io_util.cc:236]: Note to self: we could instead copy the code from Python's `errmap.h` file.

File: cpp/src/arrow/util/io_util.cc [hunk 2]
```diff
@@ -230,6 +230,150 @@ class ErrnoDetail : public StatusDetail {
...
+    case ERROR_WRONG_DISK:               //   34
+    case ERROR_SHARING_BUFFER_EXCEEDED:  //   36
+    case ERROR_NETWORK_ACCESS_DENIED:    //   65
+    case ERROR_CANNOT_MAKE:              //   82
+    case ERROR_FAIL_I24:                 //   83
+    case ERROR_DRIVE_LOCKED:             //  108
+    case ERROR_SEEK_ON_DEVICE:           //  132
+    case ERROR_NOT_LOCKED:               //  158
+    case ERROR_LOCK_FAILED:              //  167
+    case 35:                             //   35 (undefined)
```
Comment (westonpace) [inline cpp/src/arrow/util/io_util.cc:337]: I suppose this is somewhat "vendored" code but why is "35 (undefined)" fall into the category EACCES?

**Review by westonpace — APPROVED** [review]
Looks good.  Only thought I had is that there are not a lot of places remaining where we use `ARROW_GCS_RETURN_NOT_OK` and now there are some places that call `ARROW_GCS_RETURN_NOT_OK(...)` and some that call `RETURN_NOT_OK(WrapReadObjectStatus(...))` and it isn't immediately obvious why you would use one or the other (presumably you would use the latter if the status had a chance of being "file not found")

Should we just change `ARROW_GCS_RETURN_NOT_OK` to call `WrapReadObjectStatus`?

File: cpp/src/arrow/util/io_util.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/io_util.cc:337]: The legacy version of that code was generated by calling the `_dosmaperr` function, which apparently isn't supported anymore. That function did return a mapping for the Windows error code 35, which doesn't have a corresponding macro.

**Comment by pitrou** [comment]
> Should we just change `ARROW_GCS_RETURN_NOT_OK` to call `WrapReadObjectStatus`?

Not directly because `WrapReadObjectStatus` also needs a path, but we can still try to map GCS status codes to errno codes.

**Review by coryan — APPROVED** [review]
This looks cleaner, thanks.  I am afraid to ask: why emulate `errno`?  This is *not* the C API, why would the caller expect `errno` to contain anything meaningful?

**Pull request closed without merging**